docs/upgrade.rst

```diff
@@ -0,0 +1,317 @@
+----------------------------
+
+Previous pybind11 versions could only bind types globally -- all pybind11
+modules, even unrelated ones, would have access to the same exported types.
+However, this would also result in a conflict if two modules exported the
+same C++ type, which is especially problematic for very common types, e.g.
+``std::vector<int>``. :ref:`module_local` were added to resolve this (see
+that section for a complete usage guide).
+
+``py::class_`` still defaults to global bindings (because these types are
+usually unique across modules), however in order to avoid clashes of opaque
+types, ``py::bind_vector`` and ``py::bind_map`` will now bind STL containers
+as ``py::module_local`` if their elements are: builtins (``int``, ``float``,
+etc.), not bound using ``py::class_``, or bound as ``py::module_local``. For
+example, this change allows multiple modules to bind ``std::vector<int>``
+without causing conflicts. See :ref:`stl_bind` for more details.
+
+When upgrading to this version, if you have multiple modules which depend on
+a single global binding of an STL container, note that all modules can still
+accept foreign  ``py::module_local`` types in the direction of Python-to-C++.
+The locality only affects the C++-to-Python direction. If this is needed in
+multiple modules, you'll need to either:
+
+* Add a copy of the same STL binding to all of the modules which need it.
+
+* Restore the global status of that single binding by marking it
+  ``py::module_local(false)``.
+
+The latter is an easy workaround, but in the long run it would be best to
+localize all common type bindings in order to avoid conflicts with
+third-party modules.
+
+
+New syntax for custom constructors
+----------------------------------
+
+The old placement-new custom constructors are still valid, but the new approach
+greatly improves type safety (placement-new can be called accidentally with an
+incompatible type, without any compiler errors or warnings). See
+:ref:`custom_constructors` for details.
+
+.. code-block:: cpp
+
+    // old
+    py::class<Foo>(m, "Foo")
+        .def("__init__", [](Foo &self, ...) {
+            new (&self) Foo(...); // uses placement-new
+        });
+
+    // new
+    py::class<Foo>(m, "Foo")
+        .def(py::init([](...) {
+            return new Foo(...); // return by raw pointer
+            // or: return std::make_unique<Foo>(...); // return by holder
+            // or: return Foo(...); // return by value (move constructor)
+        }));
+
+
+Deprecation of some ``py::object`` APIs
+---------------------------------------
+
+The equality comparison operator for ``py::object`` only compares the
+underlying pointers, therefore ``obj1 == obj2`` in C++ is equivalent to the
+expression ``obj1 is obj2`` in Python. To avoid confusion, the equality
+operator now emits a deprecation warning (in the future, it can be replaced
+with proper rich object comparison). To compare objects by pointer, use
+``obj1.is(obj2)``.
```

Comment:
I think this should be rewritten to mention `is` first (otherwise it seems to be suggesting that `==` is the right way to do it if I don't read the paragraph in full).

```
To compare ``py::object``s by pointer, you should now use ``obj1.is(obj2)``,
which is equivalent to ``obj1 is obj2`` in Python.  Previously pybind11 used
the `==` operator for this (`obj1 == obj2`), but that could be confusing and
is now deprecated (so that it can eventually be replaced with proper rich
object comparison in a future release).
```